cmake/sca/eclair/sca.cmake

```diff
@@ -0,0 +1,140 @@
+if (CONFIG_ECLAIR_METRICS_TAB)
+  set(ECLAIR_REPORT_ADDITIONAL_OPTIONS "${ECLAIR_REPORT_ADDITIONAL_OPTIONS}" "-metrics_tab=${ECLAIR_OUTPUT_DIR}/metrics")
+endif()
+if (CONFIG_ECLAIR_REPORTS_TAB)
+  set(ECLAIR_REPORT_ADDITIONAL_OPTIONS "${ECLAIR_REPORT_ADDITIONAL_OPTIONS}" "-reports_tab=${ECLAIR_OUTPUT_DIR}/reports")
+endif()
+if (CONFIG_ECLAIR_SUMMARY_TXT)
+  set(ECLAIR_REPORT_ADDITIONAL_OPTIONS "${ECLAIR_REPORT_ADDITIONAL_OPTIONS}" "-summary_txt=${ECLAIR_OUTPUT_DIR}/summary_txt")
+endif()
+if (CONFIG_ECLAIR_SUMMARY_DOC)
+  set(ECLAIR_REPORT_ADDITIONAL_OPTIONS "${ECLAIR_REPORT_ADDITIONAL_OPTIONS}" "-summary_doc=${ECLAIR_OUTPUT_DIR}/summary_doc")
+endif()
+if (CONFIG_ECLAIR_SUMMARY_ODT)
+  set(ECLAIR_REPORT_ADDITIONAL_OPTIONS "${ECLAIR_REPORT_ADDITIONAL_OPTIONS}" "-summary_odt=${ECLAIR_OUTPUT_DIR}/summary_odt")
+endif()
+if (CONFIG_ECLAIR_FULL_TXT_ALL_AREAS)
+  set(ECLAIR_REPORT_ADDITIONAL_OPTIONS "${ECLAIR_REPORT_ADDITIONAL_OPTIONS}" "-setq=report_areas,areas")
+endif()
+if (CONFIG_ECLAIR_FULL_TXT_FIRST_AREA)
+  set(ECLAIR_REPORT_ADDITIONAL_OPTIONS "${ECLAIR_REPORT_ADDITIONAL_OPTIONS}" "-setq=report_areas,first_area")
+endif()
+if (CONFIG_ECLAIR_FULL_TXT)
+  set(ECLAIR_REPORT_ADDITIONAL_OPTIONS "${ECLAIR_REPORT_ADDITIONAL_OPTIONS}" "-full_txt=${ECLAIR_OUTPUT_DIR}/txt")
+endif()
+if (CONFIG_ECLAIR_FULL_DOC_ALL_AREAS)
+  set(ECLAIR_REPORT_ADDITIONAL_OPTIONS "${ECLAIR_REPORT_ADDITIONAL_OPTIONS}" "-setq=report_areas,areas")
+endif()
+if (CONFIG_ECLAIR_FULL_DOC_FIRST_AREA)
+  set(ECLAIR_REPORT_ADDITIONAL_OPTIONS "${ECLAIR_REPORT_ADDITIONAL_OPTIONS}" "-setq=report_areas,first_area")
+endif()
+if (CONFIG_ECLAIR_FULL_DOC)
+  set(ECLAIR_REPORT_ADDITIONAL_OPTIONS "${ECLAIR_REPORT_ADDITIONAL_OPTIONS}" "-full_doc=${ECLAIR_OUTPUT_DIR}/doc")
+endif()
+if (CONFIG_ECLAIR_FULL_ODT)
+  set(ECLAIR_REPORT_ADDITIONAL_OPTIONS "${ECLAIR_REPORT_ADDITIONAL_OPTIONS}" "-full_odt=${ECLAIR_OUTPUT_DIR}/odt")
+endif()
+
+message(STATUS "ECLAIR outputs have been written to: ${ECLAIR_OUTPUT_DIR}")
+message(STATUS "ECLAIR ECB files have been written to: ${ECLAIR_DATA_DIR}")
+message(STATUS "ECLAIR BUILD DIR is: ${ECLAIR_BUILD_DIR}")
+
+add_custom_target(eclair ALL
+  COMMAND ${CMAKE_COMMAND} -E
+    remove_directory ${ECLAIR_OUTPUT_DIR}
+  COMMAND ${CMAKE_COMMAND} -E
+    make_directory ${ECLAIR_OUTPUT_DIR}
+  COMMAND ${CMAKE_COMMAND} -E
+    make_directory ${ECLAIR_DATA_DIR}
+  COMMAND ${CMAKE_COMMAND} -E env
+    ECLAIR_DIAGNOSTICS_OUTPUT=${ECLAIR_DIAGNOSTICS_OUTPUT}
+    ECLAIR_DATA_DIR=${ECLAIR_DATA_DIR}
+    CC_ALIASES=${CC_ALIASES}
+    CXX_ALIASES=${CXX_ALIASES}
+    AS_ALIASES=${AS_ALIASES}
+    LD_ALIASES=${LD_ALIASES}
+    AR_ALIASES=${AR_ALIASES}
+    ${ECLAIR_ENV}
+    -verbose
+    -project_name=${ECLAIR_PROJECT_NAME}
+    -project_root=${ZEPHYR_BASE}
+    -eval_file=${ECLAIR_ECL_DIR}/analysis.ecl
+    -eval_file=${ECLAIR_ANALYSIS_ECL_DIR}/analysis_${ECLAIR_RULESET}.ecl
+    ${ECLAIR_ENV_ADDITIONAL_OPTIONS}
+    -- west build -p always
+                  -b ${BOARD}
```

Comment:
this doesn't look right.
In HWMv2 then `BOARD` is the board without qualifiers, so this will only work for single SoC with single core boards. Other boards will fail. (this part of the comment should be addressed)

I noticed a new `west build` invocation is done, but in a sub-dir of the existing build folder ?
I guess there is no (easy) way around it, but can you give a small explanation, and perhaps some comments in the CMake code why this is needed.

Please note that `west` is an optional (but extremely useful) tool, but that still means we generally should not rely on `west`: https://docs.zephyrproject.org/latest/develop/west/without-west.html#using-zephyr-without-west 

Can this be changed into a new CMake call, and then build can be done as `${CMAKE_COMMAND} --build <path>` ?
I notice the PR is marked for v3.7, so creation of a followup task is acceptable in this case.

Reply:
actually, this `west build` call is actually more problematic than first anticipated.

Because a user thathas set `west config build.sysbuild true` will suddenly be doing the internal eclair build using `sysbuild`, ref: https://docs.zephyrproject.org/latest/build/sysbuild/index.html#building-with-sysbuild

Also, this call doesn't seem to be able something like additional user provided build settings, for example, if user invoked the main sample build like `west build ... -- -DEXTRA_CONF_FILE=foo.conf` or ``west build ... -- -DCONFIG_BAR=y`, then this `west build` call doesn't seem to be able to handle this.

@simhein Would like your comments on this ?
Perhaps marking the Eclair support experimental and state some of the known limitations can allow us to get this in ?

Reply:
> actually, this `west build` call is actually more problematic than first anticipated.
> 
> Because a user that has set `west config build.sysbuild true` will suddenly be doing the internal eclair build using `sysbuild`, ref: [docs.zephyrproject.org/latest/build/sysbuild/index.html](https://docs.zephyrproject.org/latest/build/sysbuild/index.html#building-with-sysbuild)
> 

As far as I know and as described in documentation, they use this build call to intercept the toolchain to do the analysis stuff. It even generates a complete new output folder in the sca/EclairOutput therefore I'm unsure if the build itself is executed again. I will reach out to them to clarify it.

> Also, this call doesn't seem to be able something like additional user provided build settings, for example, if user invoked the main sample build like `west build ... -- -DEXTRA_CONF_FILE=foo.conf` or ``west build ... -- -DCONFIG_BAR=y`, then this `west build` call doesn't seem to be able to handle this.
> 

I agree, the question here is if they intercept the toolchain maybe they already have that information already but I will clarify it with them.

> @simhein Would like your comments on this ? Perhaps marking the Eclair support experimental and state some of the known limitations can allow us to get this in ?

We can mark the eclair support as experimental that would work for me.

Reply:
@tejlmand Regarding the west call and the build option without west. I was thinking is there a way to get the build command within the cmake file itself? With this I could feed the tool with what the user is calling.

Reply:
>  they use this build call to intercept the toolchain to do the analysis stuff. It even generates a complete new output folder in the sca/EclairOutput therefore I'm unsure if the build itself is executed again. I will reach out to them to clarify it.

yes, that's also my understanding, but I got a bit unsure regarding compile options / link flags / etc.
Can you clarify if those are exported from the original build and used by the tool, or if the eclair build will use the flags from the newly invoked `west build`.

> I agree, the question here is if they intercept the toolchain maybe they already have that information already but I will clarify it with them.

(y)

> We can mark the eclair support as experimental that would work for me.

Then let's do that for now.